std/mutex.zig

```diff
@@ -5,70 +5,122 @@ const AtomicRmwOp = builtin.AtomicRmwOp;
+                const rc = linux.futex_wake(&self.mutex.lock, linux.FUTEX_WAKE | linux.FUTEX_PRIVATE_FLAG, 1);
+                switch (linux.getErrno(rc)) {
+                    0 => {},
+                    linux.EINVAL => unreachable,
+                    else => unreachable,
                 }
-            } else {
-                SpinLock.Held.release(SpinLock.Held{ .spinlock = &self.mutex.spin_lock });
             }
         }
     };
 
     pub fn init() Mutex {
-        return Mutex{
-            .linux_lock = linux_lock_init,
-            .spin_lock = spin_lock_init,
+        return Mutex {
+            .lock = 0,
         };
     }
 
+    pub fn deinit(self: *Mutex) void {}
+
     pub fn acquire(self: *Mutex) Held {
-        if (builtin.os == builtin.Os.linux) {
-            var c = @cmpxchgWeak(i32, &self.linux_lock, 0, 1, AtomicOrder.Acquire, AtomicOrder.Monotonic) orelse
-                return Held{ .mutex = self };
-            if (c != 2)
-                c = @atomicRmw(i32, &self.linux_lock, AtomicRmwOp.Xchg, 2, AtomicOrder.Acquire);
-            while (c != 0) {
-                const rc = linux.futex_wait(&self.linux_lock, linux.FUTEX_WAIT | linux.FUTEX_PRIVATE_FLAG, 2, null);
-                switch (linux.getErrno(rc)) {
-                    0, linux.EINTR, linux.EAGAIN => {},
-                    linux.EINVAL => unreachable,
-                    else => unreachable,
-                }
-                c = @atomicRmw(i32, &self.linux_lock, AtomicRmwOp.Xchg, 2, AtomicOrder.Acquire);
+        var c = @cmpxchgWeak(i32, &self.lock, 0, 1, AtomicOrder.Acquire, AtomicOrder.Monotonic) orelse
+            return Held{ .mutex = self };
+        if (c != 2)
+            c = @atomicRmw(i32, &self.lock, AtomicRmwOp.Xchg, 2, AtomicOrder.Acquire);
+        while (c != 0) {
+            const rc = linux.futex_wait(&self.lock, linux.FUTEX_WAIT | linux.FUTEX_PRIVATE_FLAG, 2, null);
+            switch (linux.getErrno(rc)) {
+                0, linux.EINTR, linux.EAGAIN => {},
+                linux.EINVAL => unreachable,
+                else => unreachable,
             }
-        } else {
-            _ = self.spin_lock.acquire();
+            c = @atomicRmw(i32, &self.lock, AtomicRmwOp.Xchg, 2, AtomicOrder.Acquire);
+        }
+        return Held { .mutex = self };
+    }
+};
+
+const MutexWindows = struct {
+    lock: ?windows.RTL_CRITICAL_SECTION,
+
+    pub const Held = struct {
+        mutex: *Mutex,
+
+        pub fn release(self: Held) void {
+            windows.LeaveCriticalSection(&self.mutex.lock);
+        }
+    };
+
+    pub fn init() Mutex {
+        return Mutex {
+            .lock = null,
+        };
+    }
+
+    pub fn deinit(self: *Mutex) void {
+        windows.DeleteCriticalSection(&self.lock);
+        self.lock = null;
+    }
+
+    pub fn acquire(self: *Mutex) Held {
+        if (self.lock == null) {
+            windows.InitializeCriticalSection(&self.lock);
```

Comment:
you can't do `InitializeCriticalSection` in `init` because there is no copy elision. so the address of `lock` in `init` is different than the address of `lock` in all the other methods.

Reply:
when i move `InitializeCriticalSection` to `init` i get errors about evaluating `InitializeCriticalSection` in a constant expression because of [this](https://github.com/ziglang/zig/blob/master/std/debug/index.zig#L42) line.

Reply:
Right, I see the problem. That probably calls for the zig equivalent of `pthread_once_t`, so the mutex can be initialized on first use. I don't think we have that yet but you could emulate it with a spinlock.

(And yes, I realize the irony of suggesting a spinlock to initialize a lock.)

Reply:
yeah i thought about a spinlock but that seemed kind of redundant. i think i might close this until copy-elision lands.